File: piece_table.py
import time
class Sequence:

    def __init__(self, filepath):
        self.piece_table = PieceTable()
        self.add_buffer = ""
        #File handling code should be decoupled from the sequence so it should be in its own file
        try:
            with open(filepath, "r") as rf:
                self.file_buffer = rf.read()
        except:
            self.file_buffer = ""
        f = Span(0,0,len(self.file_buffer))
        f.next = self.piece_table.head.next
        self.piece_table.head.next = f
        f.prev = f.next.prev
        f.next.prev = f
        self.undo_stack = []
        self.redo_stack = []
        
    
    def insert(self, index, text):
        self.add_buffer += text
        offset = 0
        sptr = self.piece_table.head.next
        while sptr:
            if (offset <= index and index < offset + sptr.length):
                if offset == index:
                    new_span = Span(1, len(self.add_buffer) - len(text), len(text))
                    sptr.prev.next = new_span
                    new_span.prev = sptr.prev
                    new_span.next = sptr
                    sptr.prev = new_span
                else:
                    r_i = index-offset
                    l_span = Span(sptr.buffer, sptr.start, r_i)   
                    new_span = Span(1, len(self.add_buffer) - len(text), len(text))
                    r_span = Span(sptr.buffer, r_i + sptr.start, sptr.length - r_i)
                    sptr.prev.next = l_span
                    l_span.prev = sptr.prev
                    l_span.next = new_span
                    new_span.prev = l_span
                    new_span.next = r_span
                    r_span.prev = new_span
                    r_span.next = sptr.next
                    sptr.next.prev = r_span
                self.undo_stack.append(("erase", index, text, len(text)))
                return 1
            else:
                offset += sptr.length
                sptr = sptr.next
        return 0

    def get_text(self, buffer, start):
        if buffer == 0:
            return self.file_buffer[start]
        elif buffer == 1:
            return self.add_buffer[start]

    def erase(self, index, length):
        sptr = self.piece_table.head.next
        offset = 0
        text = ""
        l = length
        while sptr.next and length > 0:
            if (offset <= index and index < offset + sptr.length):
                if offset == index:
                    text += self.get_text(sptr.buffer, sptr.start)
                    if sptr.length == 1:
                        sptr.prev.next = sptr.next
                        sptr.next.prev = sptr.prev
                        sptr = sptr.next
                        length -= 1
                    else:
                        sptr.start += 1
                        sptr.length -= 1
                        length -= 1
                elif offset + sptr.length - 1 == index:
                    text += self.get_text(sptr.buffer, sptr.length-1)
                    sptr.length -= 1
                    sptr = sptr.next
                    length -= 1
                    offset += index
                else:
                    r_i = index-offset
                    l_span = Span(sptr.buffer, sptr.start, r_i)   
                    r_span = Span(sptr.buffer, r_i+sptr.start+1, sptr.length - r_i)
                    text += self.get_text(sptr.buffer, r_i + sptr.start)
                    sptr.prev.next = l_span
                    l_span.prev = sptr.prev
                    l_span.next = r_span
                    r_span.prev = l_span
                    r_span.next = sptr.next
                    sptr.next.prev = r_span#split span in two
                    sptr = r_span
                    length -= 1
                    offset += r_i
            else:
                offset += sptr.length
                sptr = sptr.next
        self.undo_stack.append(("insert", index, text, l))
        return 0

    def undo_redo(self, stack): #revert?
        if stack == "undo":
            pop_stack = self.undo_stack
            push_stack = self.redo_stack
        elif stack == "redo":
            push_stack = self.undo_stack
            pop_stack = self.redo_stack
        else:
            return 0
        if len(pop_stack) > 0:
            action = pop_stack.pop()
            if action[0] == "insert":
                self.insert(action[1], action[2]) 
                self.undo_stack.pop()
                #lock the linked list?
                push_stack.append(("erase", action[1], action[2], action[3]))
            elif action[0] == "erase":
                self.erase(action[1], action[3])
                self.undo_stack.pop()
                push_stack.append(("insert", action[1], action[2], action[3]))
            else:
                return 0
            return 1

    def refresh(self):
        self.file_buffer = str(self)
        self.add_buffer = ""
        f = Span(0,0,len(self.file_buffer))
        f.next = self.piece_table.tail
        self.piece_table.head.next = f
        f.prev = self.piece_table.head
        self.piece_table.tail.prev = f
  
 
         


    def __str__(self):
        output = ""
        sptr = self.piece_table.head.next
        while sptr.next:
            if not sptr.buffer:
                output += self.file_buffer[sptr.start:sptr.start+sptr.length]
            else:
                output += self.add_buffer[sptr.start:sptr.start+sptr.length]
            sptr = sptr.next
        return output
            




class PieceTable:
    
    def __init__(self):
        self.head = Span(0,0,0)
        self.tail = Span(0,0,1)
        self.head.next = self.tail
        self.tail.prev = self.head

    def __str__(self):
        sptr = self.head.next
        output = ""
        while sptr.next:
            output += str(sptr) + "\n"
            sptr = sptr.next
        return output
            


class Span:

    def __init__(self, buffer, start, length):
        self.next = None
        self.prev = None
        self.buffer = buffer
        self.start = start
        self.length = length

    def copy(self):
        new = Span(self.buffer, self.start, self.length)
        new.next = self.next
        new.prev = self.prev
        return new

    def __str__(self):
       return f"({self.buffer},{self.start},{self.length})"







s = Sequence("test.tx")
#print(len(s.file_buffer))
#s.insert(2,"!!")
#s.undo()
#s.insert(2,"yyyy")
#s.insert(6,"?",3)
#s.insert(0,"?",3)

#s.erase(1,2)
#s.erase(5,10)
#s.undo()
#s.insert(0,"xxxx")
#s.insert(2,"4")
#s.insert(0,"5")
#s.undo()
#s.undo()
#s.redo()
#s.redo()
#s.erase(1,2)
#s.erase(0,1)
#s.insert(5,"x")
def main():
    while True:
        command = input().split(" ")
        if command[0] == "insert":
            s.insert(int(command[1]), command[2])
        if command[0] == "erase":
            s.erase(int(command[1]), int(command[2]))
        if command[0] == "undo":
            s.undo_redo("undo")
        if command[0] == "redo":
            s.undo_redo("redo")
        if command[0] == "refresh":
            s.refresh()
        
        print(s.piece_table)
        print(s)
if __name__ == "__main__":
    main()

#print(s.piece_table)
#print(s.add_buffer)
s = Sequence("tet.txt")
#rf = open("test.txt","r")
#a = rf.read()
#rf.close()
#s.buffer = a
#print(len(s.buffer))
#s.insert(0,  "abcdefghigklmnopkrstuvwxyz")
st = time.time()
s.insert(22222,"1234")
#s.refresh()
en = time.time()
print(en-st)
print(s.piece_table)
#print(s)

File: src/piece_table.cpp
#include "piece_table.h"

Span::Span(bool buffer, size_t start, size_t length)
{
	this->buffer = buffer;
	this->start = start;
	this->length = length;
	this->next = nullptr;
	this->prev = nullptr;
}

//We don't want a destructor for spanranges, as some of them are stack allocated to help with some functions, and thus the destructor would free the spans inside of them when they went out of scope (memory management is complicated)
void SpanRange::free()
{
	if(!boundary){
		Span *sptr, *next, *term;
		for(sptr = first, term = last->next; sptr && sptr != term; sptr = next){
			next = sptr->next;
			delete sptr;
		}
	}
}

void SpanRange::append(Span *span)
{
	if(!first){
		first = span;
	}
	else if(first){
		last->next = span;
		span->prev  = last;
	}
	last = span;
}

void SpanRange::span_boundary(Span *before, Span *after)
{
	first = before;
	last = after;
	boundary = true;
}


PieceTable::PieceTable(std::string s)
{
	last_insert = -1;
	head = new Span(0, 0, 0);
	tail = new Span(0, 0, 1);
	head->next = tail;
	tail->prev = head;
	file_buffer = s;
	length = s.length();
	Span *f = new Span(0, 0, length);
	f->next = tail;
	f->prev = head;
	head->next = f;
	tail->prev = f;
}

PieceTable::~PieceTable()
{
	Span *sptr, *tmp;
	for(sptr = head->next; sptr->next; sptr = tmp){
		tmp = sptr->next;
		delete sptr;
	}
	delete head;
	delete tail;
	clearstack(undostack);
	clearstack(redostack);
}

void PieceTable::clearstack(std::stack<SpanRange*> &s)
{
	while(!s.empty()){
		//pop doesn't return the value for whatever reason (shit language)
		SpanRange *range = s.top();
		s.pop();
		range->free();
		delete range;
	}
}

void PieceTable::swaprange(SpanRange *src, SpanRange *dest)
{
	if(src->boundary){
		src->first->next = dest->first;
		dest->first->prev = src->first;
		src->last->prev = dest->first;
		dest->first->next = src->last;
	}
	else{
		dest->first->prev = src->first->prev;
		src->first->prev->next = dest->first;
		src->last->next->prev = dest->last;
		dest->last->next = src->last->next;
	}
}

char PieceTable::get_char_at(size_t index)
{
	Span *sptr;
	size_t offset = 0;
	for(sptr = head->next; sptr->next; sptr = sptr->next){
		if(offset <= index && index < offset + sptr->length){
			size_t r_i = index - offset;
			if(!sptr->buffer){
				return file_buffer[sptr->start + r_i];
			}
			else{
				return add_buffer[sptr->start + r_i];
			}
		}
		offset += sptr->length;
	}
	return 'a';
}



		


void PieceTable::insert(size_t index, const char *text, size_t size)
{
	SpanRange *old_span = new SpanRange();
	old_span->length = length;
	SpanRange new_span;
	for(size_t i = 0; i < size; i++){
		add_buffer.push_back(text[i]);
	}
	size_t offset = 0;
	Span *sptr;
	for(sptr = head->next; sptr; sptr = sptr->next){
		if(offset <= index && index < offset + sptr->length){
			if(offset == index){
				if(last_insert == index){
					sptr->prev->length += size;
				}
				else{
					old_span->span_boundary(sptr->prev, sptr);
					new_span.append(new Span(true, add_buffer.size() - size, size));
					swaprange(old_span, &new_span);
					undostack.push(old_span);

				}
			}
			else{
				size_t r_i = index - offset;
				old_span->append(sptr);
				new_span.append(new Span(sptr->buffer, sptr->start, r_i));
				new_span.append(new Span(true, add_buffer.size() - size, size));
				new_span.append(new Span(sptr->buffer, r_i + sptr->start, sptr->length - r_i));
				swaprange(old_span, &new_span);
				undostack.push(old_span);
			}
			length += size;
			last_insert = index + size;
			return;
		}
		else{
			offset += sptr->length;
		}
	}
}


void PieceTable::erase(size_t index, size_t size)
{
	Span *sptr;
	SpanRange *old_span = new SpanRange();
	old_span->length = length;
	SpanRange new_span;
	size_t offset = 0;
	size_t temp_size = size;
	for(sptr = head->next;sptr->next; sptr = sptr->next){
		if(offset <= index && index < offset + sptr->length){
			size_t r_i = index - offset;
			if(r_i != 0){
				new_span.append(new Span(sptr->buffer, sptr->start, r_i));
				if(r_i + size < sptr->length){
					new_span.append(new Span(sptr->buffer, sptr->start + r_i + size, sptr->length - r_i - size));
				}
				size -= std::min(size, sptr->length - r_i);
				old_span->append(sptr);
				sptr = sptr->next;
			}
			break;
		}
		offset += sptr->length;
	}
	for(; sptr->next && size > 0; sptr = sptr->next){
		if(size < sptr->length){
			new_span.append(new Span(sptr->buffer, sptr->start + size, sptr->length - size));
		}
		size -= std::min(size, sptr->length);
		old_span->append(sptr);
	}
	if(new_span.first){
		swaprange(old_span, &new_span);
	}
	else{
		old_span->first->prev->next = old_span->last->next;
		old_span->last->next->prev = old_span->first->prev;
	}
	undostack.push(old_span);
	length -= temp_size;
	last_insert = -1;
}

void PieceTable::undo_redo(int action)
{
	std::stack<SpanRange*> *src_stack;
	std::stack<SpanRange*> *dest_stack;
	if(action == 0){
		src_stack = &undostack;
		dest_stack = &redostack;
	}
	else if(action == 1){
		src_stack = &redostack;
		dest_stack = &undostack;
	}
	else{
		return;
	}
	if(src_stack->empty()){
		std::cout << "jj" << std::endl;
		return;
	}
	SpanRange *range = src_stack->top();
	size_t temp_length = length;
	length = range->length;
	range->length = temp_length;
	src_stack->pop();
	dest_stack->push(range);
	Span *first, *last;
	if(range->boundary){
		first = range->first->next;
		last = range->last->prev;
		//link the old range
		range->first->next = range->last;
		range->last->prev = range->first;
		//Store the span range
		range->first = first;
		range->last = last;
		range->boundary = false;
	}
	else{
		first = range->first->prev;
		last = range->last->next;
		if(first->next == last){
			first->next = range->first;
			last->prev = range->last;
			range->first = first;
			range->last = last;
			range->boundary = true;
		}
		else{
			first = first->next;
			last = last->prev;
			first->prev->next = range->first;
			last->next->prev = range->last;
			range->first = first;
			range->last = last;
			range->boundary = false;
		}
	}
}

std::string PieceTable::toString()
{
	std::string output;
	Span *sptr;
	for(sptr = head->next; sptr->next;sptr = sptr->next){
		if(!sptr->buffer){
			for(size_t i = sptr->start; i < sptr->start + sptr->length; i++){
				output += file_buffer[i];
			}
		}
		else{
			for(size_t i = sptr->start; i < sptr->start + sptr->length; i++){
				output += add_buffer[i];
			}
		}
	}
	return output;
}

